File: tests/test-minecraft-osp/src/main/java/pize/tests/voxelgame/main/level/ChunkManagerUtils.java
package pize.tests.voxelgame.main.level;

import pize.math.vecmath.vector.Vec2f;
import pize.math.vecmath.vector.Vec3f;
import pize.tests.voxelgame.client.chunk.ClientChunk;

import static pize.tests.voxelgame.main.chunk.ChunkUtils.*;

public class ChunkManagerUtils{
    
    public static void rebuildNeighborChunks(ClientChunk chunk){
        ClientChunk neighbor;
        
        neighbor = getNeighborChunk(chunk, -1, 0);
        if(neighbor != null)
            neighbor.rebuild(false);
        
        neighbor = getNeighborChunk(chunk, -1, 1);
        if(neighbor != null)
            neighbor.rebuild(false);
        
        neighbor = getNeighborChunk(chunk, -1, -1);
        if(neighbor != null)
            neighbor.rebuild(false);
   
        neighbor = getNeighborChunk(chunk, 1, 0);
        if(neighbor != null)
            neighbor.rebuild(false);
        
        
        neighbor = getNeighborChunk(chunk, 1, 1);
        if(neighbor != null)
            neighbor.rebuild(false);
        
        neighbor = getNeighborChunk(chunk, 1, -1);
        if(neighbor != null)
            neighbor.rebuild(false);
            
        neighbor = getNeighborChunk(chunk, 0, 1);
        if(neighbor != null)
            neighbor.rebuild(false);
        
        neighbor = getNeighborChunk(chunk, 0, -1);
        if(neighbor != null)
            neighbor.rebuild(false);
    }
    
    public static void rebuildNeighborChunks(ClientChunk chunk, int lx, int lz){
        ClientChunk neighbor;
        
        if(lx == 0){
            neighbor = getNeighborChunk(chunk, -1, 0);
            if(neighbor != null)
                neighbor.rebuild(false);
            
            if(lz == SIZE_IDX){
                neighbor = getNeighborChunk(chunk, -1, 1);
                if(neighbor != null)
                    neighbor.rebuild(false);
            }else if(lz == 0 ){
                neighbor = getNeighborChunk(chunk, -1, -1);
                if(neighbor != null)
                    neighbor.rebuild(false);
            }
        }else if(lx == SIZE_IDX){
            neighbor = getNeighborChunk(chunk, 1, 0);
            if(neighbor != null)
                neighbor.rebuild(false);
            
            if(lz == SIZE_IDX){
                neighbor = getNeighborChunk(chunk, 1, 1);
                if(neighbor != null)
                    neighbor.rebuild(false);
            }else if(lz == 0 ){
                neighbor = getNeighborChunk(chunk, 1, -1);
                if(neighbor != null)
                    neighbor.rebuild(false);
            }
        }
        
        if(lz == SIZE_IDX){
            neighbor = getNeighborChunk(chunk, 0, 1);
            if(neighbor != null)
                neighbor.rebuild(false);
        }else if(lz == 0){
            neighbor = getNeighborChunk(chunk, 0, -1);
            if(neighbor != null)
                neighbor.rebuild(false);
        }
    }
    
    
    public static float distToChunk(int x, int z, Vec3f pos){
        return Vec2f.len(x - pos.x / SIZE + 0.5F, z - pos.z / SIZE + 0.5F);
    }
    
    public static float distToChunk(int x, int z, Vec2f pos){
        return Vec2f.len(x - pos.x / SIZE + 0.5F, z - pos.y / SIZE + 0.5F);
    }
    
}

File: tests/test-minecraft-osp/src/main/kotlin/pize/tests/voxelgame/main/level/ChunkManagerUtils.kt
package pize.tests.voxelgame.main.level

import pize.math.vecmath.vector.Vec2f
import pize.math.vecmath.vector.Vec2f.Companion.len
import pize.math.vecmath.vector.Vec3f
import pize.tests.voxelgame.client.chunk.ClientChunk
import pize.tests.voxelgame.main.chunk.ChunkUtils

object ChunkManagerUtils {
    fun rebuildNeighborChunks(chunk: ClientChunk?) {
        var neighbor: ClientChunk?
        neighbor = ChunkUtils.getNeighborChunk(chunk, -1, 0)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, -1, 1)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, -1, -1)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, 1, 0)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, 1, 1)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, 1, -1)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, 0, 1)
        neighbor?.rebuild(false)
        neighbor = ChunkUtils.getNeighborChunk(chunk, 0, -1)
        neighbor?.rebuild(false)
    }

    fun rebuildNeighborChunks(chunk: ClientChunk?, lx: Int, lz: Int) {
        var neighbor: ClientChunk?
        if (lx == 0) {
            neighbor = ChunkUtils.getNeighborChunk(chunk, -1, 0)
            neighbor?.rebuild(false)
            if (lz == ChunkUtils.SIZE_IDX) {
                neighbor = ChunkUtils.getNeighborChunk(chunk, -1, 1)
                neighbor?.rebuild(false)
            } else if (lz == 0) {
                neighbor = ChunkUtils.getNeighborChunk(chunk, -1, -1)
                neighbor?.rebuild(false)
            }
        } else if (lx == ChunkUtils.SIZE_IDX) {
            neighbor = ChunkUtils.getNeighborChunk(chunk, 1, 0)
            neighbor?.rebuild(false)
            if (lz == ChunkUtils.SIZE_IDX) {
                neighbor = ChunkUtils.getNeighborChunk(chunk, 1, 1)
                neighbor?.rebuild(false)
            } else if (lz == 0) {
                neighbor = ChunkUtils.getNeighborChunk(chunk, 1, -1)
                neighbor?.rebuild(false)
            }
        }
        if (lz == ChunkUtils.SIZE_IDX) {
            neighbor = ChunkUtils.getNeighborChunk(chunk, 0, 1)
            neighbor?.rebuild(false)
        } else if (lz == 0) {
            neighbor = ChunkUtils.getNeighborChunk(chunk, 0, -1)
            neighbor?.rebuild(false)
        }
    }

    fun distToChunk(x: Int, z: Int, pos: Vec3f): Float {
        return len((x - pos.x / ChunkUtils.SIZE + 0.5f).toDouble(), (z - pos.z / ChunkUtils.SIZE + 0.5f).toDouble())
    }

    fun distToChunk(x: Int, z: Int, pos: Vec2f?): Float {
        return len((x - pos!!.x / ChunkUtils.SIZE + 0.5f).toDouble(), (z - pos.y / ChunkUtils.SIZE + 0.5f).toDouble())
    }
}
